File: src/visualizer/Turner.java
package visualizer;

import simulation.Controller;

import java.util.concurrent.atomic.AtomicBoolean;

public class Turner implements Runnable {
    volatile AtomicBoolean isAskedToStop = new AtomicBoolean(false);
    private int turns;
    private Controller controller;
    private int printTurnDelta = 50;

    public Turner(int turns, Controller controller) {
        this.turns = turns;
        this.controller = controller;
    }

    @Override
    public void run() {
        for (int i = 0; i < turns; i++) {
            controller.turn();
            if (i % printTurnDelta == 0) {
                System.out.println(controller.world.getTurn());
            }
            if (isAskedToStop.get()) {
                System.out.println("Terminating Turner thread");
                return;
            }
        }
    }
}

File: EcosystemSimulation/src/shmp/visualizer/Turner.kt
package shmp.visualizer

import shmp.simulation.Controller
import shmp.simulation.World
import java.util.concurrent.atomic.AtomicBoolean


class Turner(
        private val turns: Int,
        private val printTurnStep: Int,
        private val controller: Controller<out World>
) : Runnable {
    @JvmField
    @Volatile
    var isAskedToStop = AtomicBoolean(false)

    override fun run() {
        for (i in 0 until turns) {
            controller.turn()
            if (i % printTurnStep == 0)
                Controller.visualizer.print()

            if (isAskedToStop.get()) {
                println("Terminating Turner thread")
                return
            }
        }
    }
}
